nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -877,6 +878,120 @@ impl<'a> Planner<'a> {
     }
 
+    /// Update at most one existing zone to use a new image source.
+    fn do_plan_zone_updates(&mut self) -> Result<(), Error> {
+        // We are only interested in non-decommissioned sleds.
+        let sleds = self
+            .input
+            .all_sleds(SledFilter::Commissioned)
+            .map(|(id, _details)| id)
+            .collect::<Vec<_>>();
+
+        // Wait for all current zones to appear in the inventory.
+        // It would be nice if we could check their image source,
+        // but the inventory doesn't report that correctly: see
+        // <https://github.com/oxidecomputer/omicron/issues/8084>.
+        let inventory_zones = self
+            .inventory
+            .all_omicron_zones()
+            .map(|z| (z.id, z))
+            .collect::<BTreeMap<_, _>>();
+        for sled_id in sleds.iter().cloned() {
+            if !self
+                .blueprint
+                .current_sled_zones(
+                    sled_id,
+                    BlueprintZoneDisposition::is_in_service,
+                )
+                .all(|zone| inventory_zones.contains_key(&zone.id))
+            {
+                info!(
+                    self.log, "zones not yet up-to-date";
+                    "sled_id" => %sled_id,
+                );
+                return Ok(());
+            }
+        }
+
+        // Choose among the out-of-date zones one with the fewest dependencies.
+        let mut out_of_date_zones = sleds
+            .into_iter()
+            .flat_map(|sled_id| {
+                self.blueprint
+                    .current_sled_zones(
+                        sled_id,
+                        BlueprintZoneDisposition::is_in_service,
+                    )
+                    .filter_map(|zone| {
+                        (zone.image_source
+                            != self
+                                .blueprint
+                                .zone_image_source(zone.zone_type.kind()))
+                        .then(|| (sled_id, zone.clone()))
+                    })
+                    .collect::<Vec<_>>()
+            })
+            .collect::<Vec<(SledUuid, BlueprintZoneConfig)>>();
+
+        sort_zones_by_deps(&mut out_of_date_zones);
+        if let Some((sled_id, zone)) = out_of_date_zones.first() {
+            return self.update_or_expunge_zone(*sled_id, zone);
+        }
+
+        info!(self.log, "all zones up-to-date");
+        Ok(())
+    }
+
+    /// Update a zone to use a new image source, either in-place or by
+    /// expunging it and letting it be replaced in a future iteration.
+    fn update_or_expunge_zone(
+        &mut self,
+        sled_id: SledUuid,
+        zone: &BlueprintZoneConfig,
+    ) -> Result<(), Error> {
+        let zone_kind = zone.zone_type.kind();
+        let image_source = self.blueprint.zone_image_source(zone_kind);
+        if zone.image_source == image_source {
+            // This should only happen in the event of a planning error above.
```

Comment:
Can we return an `Err` in this case? If the planner has internal inconsistencies, we should emit something that makes us investigate.

Reply:
Indeed! 7f1f52244 logs and returns an error.